File: LeetCode/SearchInsertPosition.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace LeetCode
{
    public static class SearchInsertPosition
    {
        public static int SearchInsert(int[] nums, int target)
        {
            if (target > nums[nums.Length - 1]) return nums.Length;
            if (target < nums[0]) return 1;

            long min = 0;
            long max = nums.Length-1;
            var direction = 0; 
            while (min <= max)
            {
                long mid = (min + max) / 2;

                if (nums[mid] == target || (nums[mid] < target && direction == -1 ) || (nums[mid] > target && direction == 1))
                {
                    return (int)mid+1;
                }
                if (nums[mid]< target)
                {
                    direction= 1;
                    min = mid + 1;
                }
                else
                {
                    direction = -1;
                    max = mid - 1;
                }
            }

            return 1;
        }
    }
}

File: javascript/LeetCode/SearchInsertPosition.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */

var searchInsert = function (nums, target) {
  var l = 0;
  var r = nums.length - 1;
  if (nums[0] > target) return 0;
  if (nums[r] < target) return r + 1;
  while (l <= r) {
    var mid = l + Math.floor((r - l) / 2);

    if (nums[mid] === target) {
      return mid;
    }

    if (nums[mid] < target && nums[mid + 1] > target) {
      return mid + 1;

    }

    if (nums[mid] < target) {
      l = mid + 1;
    }
    else {
      r = mid - 1;
    }
  }
};